### PR TITLE
Matched Video Interface (vi) files and modes

### PR DESCRIPTION
**Pull request by Mr-Wiseguy**

### CODE REVIEW

File: src/io/vifade.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#include "viint.h"
+
+// TODO: this comes from a header
+#ident "$Revision: 1.17 $"
+
+void osViFade(u8 active, u16 factor) {
+    register u32 saveMask = __osDisableInt();
+
+    if (active) {
+        __osViNext->y.offset = factor & VI_2_10_FPART_MASK; //TODO: magic constant
```
Comment (Thar0) [inline src/io/vifade.c:11]: Is this TODO still relevant?

File: src/io/viint.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+//TODO: figure out what this is
+#define VI_STATE_01 0x01
+#define VI_STATE_XSCALE_UPDATED 0x02
+#define VI_STATE_YSCALE_UPDATED 0x04
+#define VI_STATE_08 0x08         //related to control regs changing
+#define VI_STATE_10 0x10         //swap buffer
+#define VI_STATE_BLACK 0x20      //probably related to a black screen
+#define VI_STATE_REPEATLINE 0x40 //repeat line?
+#define VI_STATE_FADE 0x80       //fade
```
Comment (Thar0) [inline src/io/viint.h:17]:
```suggestion
#define VI_STATE_MODE_UPDATED 0x01
#define VI_STATE_XSCALE_UPDATED 0x02
#define VI_STATE_YSCALE_UPDATED 0x04
#define VI_STATE_CTRL_UPDATED 0x08   // related to control regs changing
#define VI_STATE_BUFFER_UPDATED 0x10 // swap buffer
#define VI_STATE_BLACK 0x20          // probably related to a black screen
#define VI_STATE_REPEATLINE 0x40     // repeat line?
#define VI_STATE_FADE 0x80           // fade
```
Some suggestions for the unnamed states, optional

File: src/io/vimgr.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+void osCreateViManager(OSPri pri)
+{
+    u32 savedMask;
+    OSPri myPri;
+    OSPri oldPri;
+
+    if (!__osViDevMgr.active) {
+        __osTimerServicesInit();
+        __additional_scanline = 0;
+        osCreateMesgQueue(&viEventQueue, viEventBuf, 5);
```
Comment (Thar0) [inline src/io/vimgr.c:26]:
```suggestion
        osCreateMesgQueue(&viEventQueue, viEventBuf, ARRLEN(viEventBuf));
```

File: src/io/vimgr.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        oldPri = -1;
+        myPri = osGetThreadPri(NULL);
+
+        if (myPri < pri) {
+            oldPri = myPri;
+            osSetThreadPri(NULL, pri);
+        }
+
+        savedMask = __osDisableInt();
+        __osViDevMgr.active = 1;
```
Comment (Thar0) [inline src/io/vimgr.c:44]:
```suggestion
        __osViDevMgr.active = TRUE;
```

File: src/vimodes/vimodefpalhaf1.c [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#include "PR/os.h"
+#include "PR/rcp.h"
+#include "../io/viint.h"
+
+OSViMode osViModeFpalHaf1 = {
+    OS_VI_FPAL_HAF1,  // type
+    { // comRegs
+        VI_CTRL_TYPE_16 | VI_CTRL_GAMMA_DITHER_ON | VI_CTRL_GAMMA_ON |
+            VI_CTRL_DIVOT_ON | VI_CTRL_SERRATE_ON | 0x3000,  // ctrl
```
Comment (Thar0) [inline src/vimodes/vimodefpalhaf1.c:9]:
```suggestion
            VI_CTRL_DIVOT_ON | VI_CTRL_SERRATE_ON | VI_CTRL_PIXEL_ADV_3,  // ctrl
```
(also in other individual vimode files)
This constant isn't actually standard in libultra headers, I read about it being related to "Pixel Advance" in https://dragonminded.com/n64dev/Reality%20Coprocessor.pdf
I think it's preferable over a constant whatever the case

**Review by Thar0 — CHANGES_REQUESTED** [review]
Couple of small observations

File: src/io/vifade.c (hunk 1 above)
Comment (Mr-Wiseguy) [reply src/io/vifade.c]: Nope, removed

**Review by Thar0 — APPROVED** [review]

**Pull request merged**